pkg/ipamd/ipamd.go

```diff
@@ -523,21 +535,32 @@ func (c *IPAMContext) nodeInit() error {
 		}, 30*time.Second)
 	}
 
-	// Make a k8s client request for the current node so that max pods can be derived
-	node, err := k8sapi.GetNode(ctx, c.k8sClient)
-	if err != nil {
-		log.Errorf("Failed to get node", err)
-		podENIErrInc("nodeInit")
-		return err
-	}
-
-	maxPods, isInt64 := node.Status.Capacity.Pods().AsInt64()
-	if !isInt64 {
-		log.Errorf("Failed to parse max pods: %s", node.Status.Capacity.Pods().String)
-		podENIErrInc("nodeInit")
-		return errors.New("error while trying to determine max pods")
+	// if apiserver is connected, get the maxPods from node
+	var node corev1.Node
+	if c.withApiServer {
+		node, err := k8sapi.GetNode(ctx, c.k8sClient)
+		if err != nil {
+			log.Errorf("Failed to get node, %s", err)
+			podENIErrInc("nodeInit")
+			return err
+		} else {
+			maxPods, isInt64 := node.Status.Capacity.Pods().AsInt64()
+			if !isInt64 {
+				log.Errorf("Failed to parse max pods: %s", node.Status.Capacity.Pods().String)
+				podENIErrInc("nodeInit")
+				return errors.New("error while trying to determine max pods")
+			}
+			c.maxPods = int(maxPods)
+		}
+	} else {
+		maxPods, err := c.getMaxPodsFromFile()
```

Comment:
How will this work if we use PD ? 
Won't it mean we can schedule more Pods on the node ?

Reply:
Synced offline. During init we wait 30s for apiserver connectivity, during which most of the cases should be connected, so expect minimal impact on existing cx.
Without API server, IPAMD will run in a limited mode, meaning no k8s cache, and reading maxPods from static file. Another option is to copy kubelet config to the container and read the maxPods from there, but that may not be accurate either, and may require security review.
Once the apiserver connected in the background, we will re-read maxPods from node status, and re-create k8s cache for accuracy.